homeassistant/components/tts/__init__.py

```diff
@@ -0,0 +1,363 @@
+        filename = "{}.{}".format(key, extension)
+        voice_file = os.path.join(provider.cache_dir, filename)
+
+        if data is None:
+            raise HomeAssistantError("No TTS from %s for '%s'",
+                                     engine, message)
+
+        # save to memory
+        self._async_store_to_memcache(key, filename, data)
+
+        if cache:
+            def save_speech():
+                """Store speech to filesystem."""
+                with open(voice_file, 'wb') as speech:
+                    speech.write(data)
+
+            try:
+                yield from self.hass.loop.run_in_executor(None, save_speech)
+                self.file_cache[key] = filename
+            except OSError:
+                raise HomeAssistantError("Can't write %s", filename)
+
+        return filename
+
+    @asyncio.coroutine
+    def async_file_to_mem(self, engine, key):
+        """Load voice from file cache into memory.
+
+        This method is a coroutine.
+        """
+        provider = self.providers[engine]
+        filename = self.file_cache[key]
+
+        voice_file = os.path.join(provider.cache_dir, filename)
+
+        def load_speech():
+            """Load a speech from filesystem."""
+            with open(voice_file, 'rb') as speech:
+                return speech.read()
+
+        try:
+            data = yield from self.hass.loop.run_in_executor(None, load_speech)
+        except OSError:
+            raise HomeAssistantError("Can't read %s", voice_file)
+
+        self._async_store_to_memcache(key, filename, data)
+
+    @callback
+    def _async_store_to_memcache(self, key, filename, data):
+        """Store data to memcache and set timmer to remove it."""
+        self.mem_cache[key] = {
+            MEM_CACHE_FILENAME: filename,
+            MEM_CACHE_VOICE: data,
+        }
+
+        @callback
+        def async_remove_from_mem():
+            """Cleanup memcache."""
+            self.mem_cache.pop(key)
+
+        self.hass.loop.call_later(MEM_CACHE_TIME, async_remove_from_mem)
+
+    @asyncio.coroutine
+    def async_read_tts(self, filename):
+        """Read a voice file and return binary.
+
+        This method is a coroutine.
+        """
+        record = _RE_VOICE_FILE.match(filename)
+        if not record:
+            raise HomeAssistantError("Wrong tts file format!")
+
+        key = "{}_{}".format(record[0], record[1])
+        engine = record[1]
+
+        if key not in self.mem_cache:
+            yield from self.async_file_to_mem(engine, key)
+
+        content, _ = mimetypes.guess_type(filename)
+        return (content, self.mem_cache[key][MEM_CACHE_VOICE])
+
+
+class Provider(object):
+    """Represent a single provider."""
+
+    hass = None
+    use_cache = DEFAULT_CACHE
+    cache_dir = DEFAULT_CACHE_DIR
+    language = DEFAULT_LANG
+
+    def get_tts_audio(self, message):
+        """Load tts audio file from provider."""
+        raise NotImplementedError()
```

Comment:
Is not needed while google platform are async. Work like the other async platform entities.